src/main/scala/org/kframework/minikore/MiniKoreInterface.scala

```diff
@@ -0,0 +1,463 @@
+    }
+  }
+
+  trait NodeBuilder[T <: AST[T]] {
+    def apply(children: Seq[_ <: T]): T
+  }
+
+  trait Node0Builder[T <: AST[T]] extends NodeBuilder[T] {
+
+    def apply(): T
+
+    override def apply(children: Seq[_ <: T]) = {
+      assert(children.isEmpty)
+      apply()
+
+    }
+  }
+
+  trait Node1Builder[T <: AST[T]] extends NodeBuilder[T] {
+
+    def apply(p: T): T
+
+    override def apply(children: Seq[_ <: T]) = {
+      assert(children.size == 1)
+      apply(children.head)
+    }
+  }
+
+
+  trait Node2Builder[T <: AST[T]] extends NodeBuilder[T] {
+
+    def apply(p: T, q: T): T
+
+    override def apply(children: Seq[_ <: T]) = {
+      assert(children.size == 2)
+      apply(children.head, children(1))
+    }
+  }
+
+
+  sealed trait Leaf[T <: AST[T], C] extends AST[T] {
+    def contents: C
+
+    def build: LeafBuilder[T, C]
+  }
+
+  object Leaf {
+    def unapply[T <: AST[T], C](arg: AST[T]): Option[C] = arg match {
+      case l: Leaf[T, C] => Some(l.contents)
+      case _ => None
+    }
+  }
+
+  trait LeafBuilder[T <: AST[T], C] {
+    def apply(contents: C): T
+  }
+
+
+  trait LabelledNode[L, T <: AST[T]] extends Node[T] {
+    def label: L
+  }
+
+  trait LabelledNodeBuilder[L, T <: AST[T]] {
+    def apply(x: L, c: Seq[_ <: T]): T
+  }
+
+  object LabelledNode {
+    def unapply[L, T <: AST[T]](arg: AST[T]): Option[(L, Seq[_ <: T])] = arg match {
+      case p: LabelledNode[L, T] => Some(p.label, p.children)
+      case _ => None
+    }
+  }
+
+
+  sealed trait Node0[T <: AST[T]] extends Node[T] {
+    override def children = Seq.empty
+
+    def build: Node0Builder[T]
+
+  }
+
+  object Node0 {
+    def unapply[T <: AST[T]](arg: AST[T]): Boolean = arg match {
+      case p: Node0[T] => true
+      case _ => false
+    }
+  }
+
+  sealed trait Node1[T <: AST[T]] extends Node[T] with Product1[T] {
+
+    def build: NodeBuilder[T]
+
+    override def children = Seq(_1)
+  }
+
+  object Node1 {
+    def unapply[T <: AST[T]](arg: AST[T]): Option[T] = arg match {
+      case n: Node1[T] => Some(n._1)
+      case _ => None
+    }
+  }
+
+
+  sealed trait Node2[T <: AST[T]] extends Node[T] with Product2[T, T] {
+
+    override def children = Seq(_1, _2)
+
+    def build: Node2Builder[T]
+
+  }
+
+  object Node2 {
+    def unapply[T <: AST[T]](arg: AST[T]): Option[(T, T)] = arg match {
+      case n: Node2[T] => Some((n._1, n._2))
+      case _ => None
+    }
+  }
+
+}
+
+object PatternInterface {
+
+  import TreeInterface._
+
+  sealed trait Pattern extends AST[Pattern]
+
+  type Sort = String
+
+  trait Variable extends Pattern with Leaf[Pattern, (String, Sort)] {
+    def name: String
+
+    def sort: Sort
+
+    override def contents: (String, Sort) = (name, sort)
+
+  }
+
+
+  trait VariableBuilder extends LeafBuilder[Pattern, (String, Sort)] {
+
+    def apply(name: String, sort: Sort): Variable
+
+    override def apply(contents: (String, Sort)) = apply(contents._1, contents._2)
+
+  }
+
+  object Variable {
```

Comment:
We touched upon this last week. Having objects with `unapply` declared at this level will introduce an ambiguity between them and the implementations of the `Builder` objects. E.g., writing `case And(x)` will fail at compile time because `And` is declared both here and as a `Builder`.